chasm/lib/callback/executor.go

```diff
@@ -0,0 +1,219 @@
+	// directly update the callback's Status field to SUCCEEDED/BACKING_OFF/FAILED.
+	_, _, err = chasm.UpdateComponent(
+		ctx,
+		invokerRef,
+		func(c *Callback, ctx chasm.MutableContext, _ any) (struct{}, error) {
+			c.Status = result
+			c.CallbackState = callback.CallbackState
+			return struct{}{}, nil
+		},
+		nil,
+	)
+	if err != nil {
+		return fmt.Errorf("failed to update component: %w", err)
+	}
+
+	return nil
+}
+
+func (e *InvocationTaskExecutor) Validate(
+	ctx chasm.Context,
+	callback *Callback,
+	_ chasm.TaskAttributes,
+	_ *callbackspb.InvocationTask,
+) (bool, error) {
+	return callback.Status == callbackspb.CALLBACK_STATUS_SCHEDULED, nil
+}
+
+// BackoffTaskExecutor is responsible for the retry scheduling after failed
+// attempts. This is a timer executor that will fire when the backoff period
+// expires and the callback should be retried.
+type BackoffTaskExecutor struct {
+	BackoffTaskExecutorOptions
+}
+
+type BackoffTaskExecutorOptions struct {
+	fx.In
+
+	Config         *Config
+	MetricsHandler metrics.Handler
+	Logger         log.Logger
+}
+
+func NewBackoffTaskExecutor(opts BackoffTaskExecutorOptions) *BackoffTaskExecutor {
+	return &BackoffTaskExecutor{
+		BackoffTaskExecutorOptions: opts,
+	}
+}
+
+// Execute transitions the callback from BACKING_OFF to SCHEDULED state
+// and generates an InvocationTask for the next attempt. Directly
+// manipulate the state schedule the next invocation attempt.
+func (e *BackoffTaskExecutor) Execute(
+	ctx chasm.MutableContext,
+	callback *Callback,
+	_ chasm.TaskAttributes,
+	_ *callbackspb.BackoffTask,
+) error {
+	// Clear the next attempt schedule time and transition to scheduled
+	callback.NextAttemptScheduleTime = nil
+	callback.Status = callbackspb.CALLBACK_STATUS_SCHEDULED
+
+	// Generate an invocation task and task attributes for the next attempt
+	invocationTask, taskAttrs, err := e.generateInvocationTask(callback)
+	if err != nil {
+		return fmt.Errorf("failed to generate invocation task: %w", err)
+	}
+
+	// Add the task to be executed
+	ctx.AddTask(callback, taskAttrs, invocationTask)
+
+	return nil
+}
+
+func (e *BackoffTaskExecutor) Validate(
+	ctx chasm.Context,
+	callback *Callback,
+	_ chasm.TaskAttributes,
+	_ *callbackspb.BackoffTask,
+) (bool, error) {
+	// Validate that the callback is in BACKING_OFF state
+	return callback.Status == callbackspb.CALLBACK_STATUS_BACKING_OFF, nil
+}
+
+// generateInvocationTask creates an InvocationTask and TaskAttributes based on the callback variant.
+// This is the CHASM port of HSM's RegenerateTasks logic from statemachine.go:79-100.
+// The destination (scheme + host) is extracted from the URL and used for task routing,
+// matching HSM's pattern where InvocationTask stores destination as u.Scheme + "://" + u.Host.
+func (e *BackoffTaskExecutor) generateInvocationTask(callback *Callback) (*callbackspb.InvocationTask, chasm.TaskAttributes, error) {
```

Comment:
I would change this to a function that just generates the attributes. The task is easy to generate.